src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -369,11 +374,32 @@ def visit_Call(self, node: foast.Call, **kwargs) -> foast.Call:
             new_type = ct.FieldType(dims=new_dims, dtype=new_func.type.dtype)
             return foast.Call(func=new_func, args=new_args, location=node.location, type=new_type)
         elif isinstance(new_func.type, ct.FunctionType):
+            new_args = self.visit(node.args, **kwargs)
+            # todo(tehrengruber): solve in a more generic way, e.g. using
+            #  parametric polymorphism.
+            if node.func.id == "neighbor_sum":
+                field_type: ct.FieldType = new_args[0].type
+                reduction_dim = cast(ct.DimensionType, new_args[1].type).dim
+                if reduction_dim not in field_type.dims:
+                    field_dims_str = ", ".join(str(dim) for dim in field_type.dims)
+                    raise FieldOperatorTypeDeductionError.from_foast_node(
+                        node,
+                        msg=f"Incompatible field argument in {node.func.id}. "
+                        f"Expected a field with dimension {reduction_dim}, "
+                        f"but got {field_dims_str}.",
+                    )
+                return_type = ct.FieldType(
+                    dims=[dim for dim in field_type.dims if dim != reduction_dim],
+                    dtype=field_type.dtype,
+                )
```

Comment:
Might I suggest the pattern from `FieldOperatorLowering`:

```python
        if node.func.id in FUN_BUILTIN_NAMES:
            visitor = getattr(self, f"_visit_{node.func.id}")
            return visitor(node, **kwargs)
        ...
        
    def _visit_neighbor_sum(self, node: foast.Call, **kwargs):
        <function specific stuff here>
```

Reply:
1) aliasing would not be supported. It's not exactly on purpose but it is a bearable limitation for now.

2) function calls on expressions other than a `Name` lead to a syntax error before type deduction.

Reply:
I sketched how to support this with Enrique a while ago. By lifting the identity of built-in functions to the type level we could dispatch without relying on what is stored inside `Name`. If this is needed we can implement it rather quickly, but until typing on the iterator level and its relation to the frontend is resolved I am more comfortable with keeping the type system lean. Note that the aliasing limitation also applies to data types, which prevents patterns like these:
```
dtype = float64
@field_operator
def foo(...):
  return dtype("1.")
```